From aa338f7459845153ce80d89ab7210a0f4b7c9cd0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Wed, 22 Apr 2026 15:59:54 +0200
Subject: [PATCH 19/47] dinput: Check for disabled/overridden devices in
 hid_joystick_try_open.

Based on patches by Connor McAdams.

(cherry picked from commit 4770bb50e66363caa89c8415751ec6547871cebe)

CW-Bug-Id: #25883
---
 dlls/dinput/joystick_hid.c | 17 +++++++++--------
 1 file changed, 9 insertions(+), 8 deletions(-)

diff --git a/dlls/dinput/joystick_hid.c b/dlls/dinput/joystick_hid.c
index 2a83e9517ba..f9f3758ae7b 100644
--- a/dlls/dinput/joystick_hid.c
+++ b/dlls/dinput/joystick_hid.c
@@ -1508,6 +1508,7 @@ static HRESULT hid_joystick_device_try_open( const WCHAR *path, HANDLE *device,
     ULONG node_count;
     NTSTATUS status;
     UINT32 handle;
+    BOOL override;
     USHORT count;
 
     device_file = CreateFileW( path, GENERIC_READ | GENERIC_WRITE, FILE_SHARE_READ | FILE_SHARE_WRITE,
@@ -1530,6 +1531,8 @@ static HRESULT hid_joystick_device_try_open( const WCHAR *path, HANDLE *device,
     if (!HidD_GetProductString( device_file, instance->tszInstanceName, MAX_PATH * sizeof(WCHAR) )) goto failed;
     if (!HidD_GetProductString( device_file, instance->tszProductName, MAX_PATH * sizeof(WCHAR) )) goto failed;
 
+    if (device_instance_is_disabled( instance->tszInstanceName, &override )) goto failed;
+
     if (!DeviceIoControl( device_file, IOCTL_HID_GET_WINE_RAWINPUT_HANDLE, NULL, 0, &handle, sizeof(handle), &size, NULL ))
     {
         ERR( "failed to get raw input handle, error %lu\n", GetLastError() );
@@ -1647,7 +1650,7 @@ static HRESULT hid_joystick_device_try_open( const WCHAR *path, HANDLE *device,
 
     *device = device_file;
     *preparsed = preparsed_data;
-    return DI_OK;
+    return override ? S_FALSE : DI_OK;
 
 failed:
     CloseHandle( device_file );
@@ -1665,8 +1668,8 @@ static HRESULT hid_joystick_device_open( int index, const GUID *guid, DIDEVICEIN
     SP_DEVINFO_DATA devinfo = {.cbSize = sizeof(devinfo)};
     WCHAR device_id[MAX_PATH], *tmp;
     HDEVINFO set, xi_set;
-    BOOL override;
     UINT32 i = 0;
+    HRESULT hr;
     GUID hid;
 
     TRACE( "index %d, guid %s\n", index, debugstr_guid( guid ) );
@@ -1684,14 +1687,11 @@ static HRESULT hid_joystick_device_open( int index, const GUID *guid, DIDEVICEIN
         detail->cbSize = sizeof(SP_DEVICE_INTERFACE_DETAIL_DATA_W);
         if (!SetupDiGetDeviceInterfaceDetailW( set, &iface, detail, sizeof(buffer), NULL, &devinfo ))
             continue;
-        if (FAILED(hid_joystick_device_try_open( detail->DevicePath, device, preparsed,
-                                                 attrs, caps, instance, version )))
+        if (FAILED(hr = hid_joystick_device_try_open( detail->DevicePath, device, preparsed,
+                                                      attrs, caps, instance, version )))
             continue;
 
-        if (device_instance_is_disabled( instance->tszInstanceName, &override ))
-            goto next;
-
-        if (override && SetupDiGetDeviceInstanceIdW( set, &devinfo, device_id, MAX_PATH, NULL ) &&
+        if (hr == S_FALSE && SetupDiGetDeviceInstanceIdW( set, &devinfo, device_id, MAX_PATH, NULL ) &&
             (tmp = wcsstr( device_id, L"&IG_" )))
         {
             memcpy( tmp, L"&XI_", sizeof(L"&XI_") - sizeof(WCHAR) );
@@ -2142,6 +2142,7 @@ HRESULT hid_joystick_create_device( struct dinput *dinput, const GUID *guid, IDi
         wcscpy( impl->device_path, *(const WCHAR **)guid );
         hr = hid_joystick_device_try_open( impl->device_path, &impl->device, &impl->preparsed, &attrs,
                                            &impl->caps, &impl->base.instance, dinput->dwVersion );
+        if (hr == S_FALSE) hr = S_OK;
     }
     if (hr != DI_OK) goto failed;
 
-- 
2.54.0

